libs/math/include/math/tensor_broadcast.hpp

```diff
@@ -125,6 +125,47 @@ inline bool UpgradeIteratorFromBroadcast(SizeVector const &a, TensorSliceIterato
   return true;
 }
 
+template <typename T, typename C>
+inline bool UpgradeIteratorFromBroadcast(SizeVector const &              a,
```

Comment:
the code inside this UpgradeIteratorFromBroadcast method is identical to the non const version, therefore you should have an implementation function that is invoked by both to avoid code repetition.

e.g.

template <typename IteratorType>
UpgradeIteratorFromBroadcastImplementation(SizeVector const & a, IteratorType & iterator)
{
 // implementation
}